File: MB-Secretaria/providers_py/provider_response.py
from __future__ import annotations

import hashlib
from datetime import datetime, timezone
from typing import Any

from .types import ProviderRequest, ProviderResponse


def build_provider_audit_ref(request: ProviderRequest) -> str:
    payload = f"{request.trace_id}:{request.tenant_id}:{request.request_type}"
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()


def build_provider_response(
    request: ProviderRequest,
    *,
    provider_name: str = "none",
    model_name: str = "none",
    status: str = "fallback",
    content: str = "",
    latency_ms: int = 0,
    structured_output: Any = None,
    confidence_score: float | None = None,
    safety_flags: list[str] | None = None,
    fallback_used: bool = False,
    retry_count: int = 0,
) -> ProviderResponse:
    return ProviderResponse(
        trace_id=request.trace_id,
        provider_name=provider_name,
        model_name=model_name,
        latency_ms=latency_ms,
        status=status,
        content=content,
        structured_output=structured_output,
        confidence_score=confidence_score,
        safety_flags=safety_flags or [],
        fallback_used=fallback_used,
        retry_count=retry_count,
        audit_ref=build_provider_audit_ref(request),
        created_at=datetime.now(timezone.utc).isoformat(),
    )

File: MetaBrain/providers/provider-response.ts
import { createHash } from "node:crypto";
import type { ProviderName, ProviderRequest, ProviderResponse, ProviderStatus } from "./types";

export function buildProviderAuditRef(input: Pick<ProviderRequest, "trace_id" | "tenant_id" | "request_type">): string {
  return createHash("sha256").update(`${input.trace_id}:${input.tenant_id}:${input.request_type}`).digest("hex");
}

export function buildProviderResponse(input: {
  request: ProviderRequest;
  provider_name?: ProviderName | "none";
  model_name?: string;
  status: ProviderStatus;
  content?: string;
  latency_ms?: number;
  structured_output?: unknown;
  confidence_score?: number;
  safety_flags?: string[];
  fallback_used?: boolean;
  retry_count?: number;
}): ProviderResponse {
  return {
    trace_id: input.request.trace_id,
    provider_name: input.provider_name ?? "none",
    model_name: input.model_name ?? "none",
    latency_ms: input.latency_ms ?? 0,
    status: input.status,
    content: input.content ?? "",
    structured_output: input.structured_output,
    confidence_score: input.confidence_score,
    safety_flags: input.safety_flags ?? [],
    fallback_used: input.fallback_used ?? false,
    retry_count: input.retry_count ?? 0,
    audit_ref: buildProviderAuditRef(input.request),
    created_at: new Date().toISOString(),
  };
}
